compiler/semantic/scope.go

```diff
@@ -7,68 +7,52 @@ import (
 	"github.com/brimdata/zed/compiler/ast/dag"
 	"github.com/brimdata/zed/compiler/kernel"
 	"github.com/brimdata/zed/zson"
+	"golang.org/x/exp/slices"
 )
 
 type Scope struct {
-	stack []*Binder
+	parent   *Scope
+	children []*Scope
+	nvar     int
+	symbols  map[string]*entry
 }
 
-func NewScope() *Scope {
-	return &Scope{}
-}
-
-func (s *Scope) tos() *Binder {
-	return s.stack[len(s.stack)-1]
-}
-
-func (s *Scope) Enter() {
-	s.stack = append(s.stack, NewBinder())
+func NewScope(parent *Scope) *Scope {
+	s := &Scope{parent: parent, symbols: make(map[string]*entry)}
+	if parent != nil {
+		parent.children = append(parent.children, s)
+	}
+	return s
 }
 
-func (s *Scope) Exit() {
-	s.stack = s.stack[:len(s.stack)-1]
+type entry struct {
+	ref    any
+	refcnt int
```

Comment:
Is this field used?

Reply:
but I kind of think since it already exists it should be removed in a separate pr.